File: src/main/java/io/vproxy/vfx/robot/RobotWrapper.java
package io.vproxy.vfx.robot;

import io.vproxy.base.util.LogType;
import io.vproxy.base.util.Logger;
import io.vproxy.vfx.entity.input.Key;
import io.vproxy.vfx.util.FXUtils;
import javafx.embed.swing.SwingFXUtils;
import javafx.geometry.Point2D;
import javafx.scene.image.Image;
import javafx.scene.image.WritableImage;
import javafx.scene.robot.Robot;
import javafx.stage.Screen;

import java.awt.*;

public class RobotWrapper {
    private final Robot robot;
    private final java.awt.Robot awtRobot;

    public RobotWrapper() {
        this.robot = new Robot();
        java.awt.Robot awtRobot;
        try {
            awtRobot = new java.awt.Robot();
        } catch (AWTException e) {
            Logger.error(LogType.SYS_ERROR, "failed creating awt robot", e);
            awtRobot = null;
        }
        this.awtRobot = awtRobot;
    }

    public void press(Key key) {
        if (key.button != null) {
            robot.mousePress(key.button);
            assert Logger.lowLevelDebug("mouse press: " + key);
        } else if (key.key != null) {
            robot.keyPress(key.key.java);
            assert Logger.lowLevelDebug("key press: " + key);
        }
    }

    public void release(Key key) {
        if (key.button != null) {
            robot.mouseRelease(key.button);
            assert Logger.lowLevelDebug("mouse release: " + key);
        } else if (key.key != null) {
            robot.keyRelease(key.key.java);
            assert Logger.lowLevelDebug("key release: " + key);
        }
    }

    public Image captureScreen(Screen screen) {
        assert Logger.lowLevelDebug("screen capture: " + screen);
        var bounds = screen.getBounds();
        return capture0(null, bounds.getMinX(), bounds.getMinY(), (int) bounds.getWidth(), (int) bounds.getHeight(), true);
    }

    public Image capture(double x, double y, int width, int height) {
        return capture(null, x, y, width, height);
    }

    public Image capture(WritableImage img, double x, double y, int width, int height) {
        return capture(img, x, y, width, height, false);
    }

    public Image capture(WritableImage img, double x, double y, int width, int height, boolean tryAwt) {
        assert Logger.lowLevelDebug("partial capture: (" + x + ", " + y + ") + (" + width + " * " + height + ")");
        return capture0(img, x, y, width, height, tryAwt);
    }

    private Image capture0(WritableImage img, double x, double y, int width, int height, boolean tryAwt) {
        if (tryAwt && awtRobot != null) {
            var mi = awtRobot.createMultiResolutionScreenCapture(new Rectangle((int) x, (int) y, width, height));
            var ls = mi.getResolutionVariants();
            if (!ls.isEmpty()) {
                var i = ls.get(ls.size() - 1);
                boolean useAwt = false;
                if (img == null) {
                    useAwt = true;
                } else {
                    var w = i.getWidth(null);
                    var h = i.getHeight(null);
                    if (w <= img.getWidth() && h <= img.getHeight()) {
                        useAwt = true;
                    }
                }
                if (useAwt) {
                    assert Logger.lowLevelDebug("using awt captured image");
                    return SwingFXUtils.toFXImage(FXUtils.convertToBufferedImage(i), img);
                }
            }
        }
        assert Logger.lowLevelDebug("using javafx captured image");
        return robot.getScreenCapture(img, x, y, width, height);
    }

    public java.awt.Image awtCapture(int x, int y, int width, int height) {
        var mi = awtRobot.createMultiResolutionScreenCapture(new Rectangle(x, y, width, height));
        var ls = mi.getResolutionVariants();
        if (ls.isEmpty()) return null;
        return ls.get(ls.size() - 1);
    }

    public void mouseMove(double x, double y) {
        assert Logger.lowLevelDebug("mouse move: (" + x + ", " + y + ")");
        robot.mouseMove(x, y);
    }

    public Point2D getMousePosition() {
        return robot.getMousePosition();
    }

    public void mouseWheel(int wheelAmt) {
        assert Logger.lowLevelDebug("mouse wheel: " + wheelAmt);
        robot.mouseWheel(wheelAmt);
    }
}

File: src-ts/src/robot/RobotWrapper.ts
import { invoke } from '@tauri-apps/api/core';
import { Logger } from '../vproxy-base/Logger.js';
import { Key, MouseButton } from '../entity/input/Key.js';
import { MouseWheelScroll, MouseWheelDirection } from '../entity/input/MouseWheelScroll.js';

/**
 * Mirrors `io.vproxy.vfx.robot.RobotWrapper`.
 *
 * The JavaFX original drives `javafx.scene.robot.Robot` (with an `awt.Robot`
 * fallback for screen capture). The DOM cannot synthesize input or capture
 * the screen, so each operation is delegated to a native Tauri command
 * implemented in Rust (`src-tauri/src/robot.rs`) backed by the `enigo` (input
 * simulation) and `xcap` (screen capture) crates.
 *
 * Key identity is transported as the DOM `KeyboardEvent.code` of the pressed
 * key, which the Rust side maps to the native key. Mouse buttons use the
 * framework wire values "PRIMARY"/"SECONDARY"/"MIDDLE". Screen captures
 * resolve to a PNG `data:` URL (base64) since the DOM has no direct pixel
 * buffer type.
 *
 * Unlike the synchronous Java API, every method here is async (invoke returns
 * a Promise). Callers must `await` results.
 */
export class RobotWrapper {
  /**
   * Press a key/mouse-button/wheel according to which field of `key` is set,
   * matching Java's `RobotWrapper.press(Key)`.
   */
  async press(key: Key): Promise<void> {
    try {
      if (key.button !== null) {
        await invoke('robot_mouse_press', { button: key.button });
        return;
      }
      if (key.scroll !== null) {
        await invoke('robot_mouse_wheel', { amount: scrollAmount(key.scroll) });
        return;
      }
      if (key.key !== null) {
        await invoke('robot_key_press', { domCode: key.key.domCode, isLeft: key.isLeftKey });
        return;
      }
    } catch (e) {
      Logger.warn('SYS_ERROR', `RobotWrapper.press failed: ${(e as Error).message ?? e}`);
    }
  }

  /** Release a key/mouse-button, matching Java's `RobotWrapper.release(Key)`. */
  async release(key: Key): Promise<void> {
    try {
      if (key.button !== null) {
        await invoke('robot_mouse_release', { button: key.button });
        return;
      }
      if (key.key !== null) {
        await invoke('robot_key_release', { domCode: key.key.domCode, isLeft: key.isLeftKey });
        return;
      }
    } catch (e) {
      Logger.warn('SYS_ERROR', `RobotWrapper.release failed: ${(e as Error).message ?? e}`);
    }
  }

  /** Move the mouse cursor to absolute screen coordinates. */
  async mouseMove(x: number, y: number): Promise<void> {
    try {
      await invoke('robot_mouse_move', { x, y });
    } catch (e) {
      Logger.warn('SYS_ERROR', `RobotWrapper.mouseMove failed: ${(e as Error).message ?? e}`);
    }
  }

  /**
   * Scroll the vertical mouse wheel. Negative = up, positive = down (AWT
   * convention, matching Java's `Robot.mouseWheel`).
   */
  async mouseWheel(wheelAmt: number): Promise<void> {
    try {
      await invoke('robot_mouse_wheel', { amount: wheelAmt });
    } catch (e) {
      Logger.warn('SYS_ERROR', `RobotWrapper.mouseWheel failed: ${(e as Error).message ?? e}`);
    }
  }

  /** Current mouse cursor position in absolute screen pixels. */
  async getMousePosition(): Promise<{ x: number; y: number }> {
    try {
      return await invoke<{ x: number; y: number }>('robot_get_mouse_position');
    } catch (e) {
      Logger.warn('SYS_ERROR', `RobotWrapper.getMousePosition failed: ${(e as Error).message ?? e}`);
      return { x: 0, y: 0 };
    }
  }

  /** Capture the entire primary monitor as a PNG data URL. */
  async captureScreen(): Promise<string> {
    return invoke<string>('robot_capture_screen');
  }

  /**
   * Capture a sub-rectangle of the primary monitor (screen pixels, relative to
   * the monitor's top-left) as a PNG data URL.
   */
  async capture(x: number, y: number, width: number, height: number): Promise<string> {
    return invoke<string>('robot_capture', { x, y, width, height });
  }
}

/**
 * Convert a `MouseWheelScroll` to the signed AWT-style wheel amount Java's
 * `RobotWrapper` computes: magnitude defaults to 1 when 0, and UP negates the
 * value.
 */
function scrollAmount(scroll: MouseWheelScroll): number {
  let n = scroll.value;
  if (n === 0) n = 1;
  if (scroll.direction === MouseWheelDirection.UP) {
    n = -n;
  }
  return n;
}

// Re-export so callers importing from RobotWrapper still get the input types,
// matching the Java package surface.
export { Key, MouseButton, MouseWheelScroll, MouseWheelDirection };
